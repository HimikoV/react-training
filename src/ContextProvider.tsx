import { createContext, useEffect, useState, useMemo } from "react";
import { mapBasketData } from "./mappers/mapBasketData";
import { getDeliveryOptions } from "./service/getDeliveryOptions";
import { getOrderData } from "./service/getOrderData";
import { getPaymentOptions } from "./service/getPaymentOptions";
import { requestData } from "./service/requestData";
import { basketType } from "./types/basketType";
import { finalizeOptionsType } from "./types/finalizeOptionsType";
import { orderType } from "./types/orderType";
import { productType } from "./types/productType";
import { childrenInterface } from "./interfaces/childrenInterface";
import { basketProcessedType } from "./types/basketProcessedType";
import { contextProviderInterface } from "./interfaces/contextProviderInterface";

export const dataContext = createContext<contextProviderInterface>(
  {} as contextProviderInterface
);

const ContextProvider = ({ children }: childrenInterface) => {
  const [basketData, setBasketData] = useState<basketType[]>(
    JSON.parse(localStorage.getItem("basketData") || "[]")
  );
  const [filter, setFilter] = useState<string[]>([]);
  const [data, setData] = useState<productType[]>([]);
  const [loadLimit, setLoadLimit] = useState<number>(9);
  const [basketProcessedData, setBasketProcessedData] = useState<
    basketProcessedType[]
  >([]);
  const [deliveryOptions, setDeliveryOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [paymentOptions, setPaymentOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [open, setOpen] = useState<boolean>(false);
  const [orderData, setOrderData] = useState<orderType[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const responseData = await requestData();
      setData(responseData);
      const responseDelivery = await getDeliveryOptions();
      setDeliveryOptions(responseDelivery);
      const responsePayment = await getPaymentOptions();
      setPaymentOptions(responsePayment);
      const responseOrderData = await getOrderData();
      setOrderData(responseOrderData);
    };
    fetchData();
  }, []);

  useEffect(() => {
    localStorage.setItem("basketData", JSON.stringify(basketData));
  }, [basketData]);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
  }, [data]);

  const trueValues = useMemo(
    () => ({
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    }),
    [
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    ]
  );

  // const values = {};
  return (
    <dataContext.Provider value={trueValues}>{children}</dataContext.Provider>
  );
};
export default ContextProvider;
