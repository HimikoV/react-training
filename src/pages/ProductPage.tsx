import { useEffect, useState, createContext } from "react";
import ProductList from "../components/products/ProductList";
import { productType } from "../types/productType";
import ProductFilter from "../components/products/ProductFilter";
import uniqueId from "lodash.uniqueid";
import { requestData } from "../service/requestData";
import { Button } from "@mui/material";

const dataContext = createContext<productType[]>([{}] as productType[]);

const ProductPage = () => {
  const [data, setData] = useState<productType[]>([{}] as productType[]);
  const [load, setLoad] = useState<[boolean, number]>([true, 9]);
  const [order, setOrder] = useState<boolean>(false);
  const [filter, setFilter] = useState<string>("");

  const sortByLetter = (): void => {
    setOrder(!order);
    order
      ? setData(data.sort((a, b) => (a.name > b.name ? 1 : -1)))
      : setData(data.sort((a, b) => (a.name > b.name ? -1 : 1)));
  };

  const sortByPrice = (): void => {
    setOrder(!order);
    order
      ? setData(data.sort((a, b) => (a.price > b.price ? 1 : -1)))
      : setData(data.sort((a, b) => (a.price > b.price ? -1 : 1)));
  };

  const filterByName = (): void => {
    setData(data.filter((item) => item.category === filter));
  };

  const mapData = (loadedData: any): void => {
    const items: productType[] = [];
    loadedData.map((item: productType) =>
      items.push({ key: uniqueId("item_"), ...item })
    );
    setData(items);
  };

  const fetchData = async () => {
    const responseData = await requestData(load[1]);
    mapData(responseData);
    setLoad([false, load[1]]);
  };

  useEffect(() => {
    if (load[0]) {
      fetchData();
    }
    if (filter !== "") {
      filterByName();
    }
  }, [load, order, filter]);

  return (
    <>
      <ProductFilter
        sortByLetter={sortByLetter}
        sortByPrice={sortByPrice}
        setFilter={setFilter}
      />
      <dataContext.Provider value={data}>
        <ProductList />
      </dataContext.Provider>

      <Button
        variant="contained"
        onClick={() => {
          setLoad([true, load[1] + 9]);
        }}
      >
        Load More...
      </Button>
    </>
  );
};

export { ProductPage, dataContext };
