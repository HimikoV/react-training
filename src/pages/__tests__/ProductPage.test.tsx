import { ProductPage } from "../ProductPage";
import { describe, it } from "@jest/globals";
import renderer from "react-test-renderer";
import ContextProvider from "../../ContextProvider";

describe("testing ProductPage render", () => {
  it("ProductPage Renders Correctly", () => {
    const productPage = renderer
      .create(
        <ContextProvider>
          <ProductPage />
        </ContextProvider>
      )
      .toJSON();
    expect(productPage).toMatchSnapshot();
  });
});
