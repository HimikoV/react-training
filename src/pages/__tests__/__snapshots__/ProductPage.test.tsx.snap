// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ProductPage render ProductPage Renders Correctly 1`] = `
<div
  className="MuiBox-root css-1v719cu"
>
  <div
    className="MuiGrid-root css-b7ixwv-MuiGrid-root"
  >
    <img
      alt="BG for products Page"
      className="MuiBox-root css-udfay2"
      src="https://images.unsplash.com/photo-1507546530-14a03f8d180a?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=642&q=80"
    />
  </div>
  <div
    className="MuiGrid-root css-maqey9-MuiGrid-root"
  >
    <div
      className="MuiGrid-root css-er906r-MuiGrid-root"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="FilterListIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
          />
        </svg>
      </button>
    </div>
    <div
      className="MuiGrid-root css-iectcq-MuiGrid-root"
    />
  </div>
</div>
`;
